src/sound_openal.cpp

```diff
@@ -567,16 +595,55 @@ class OpenALSoundManager: public ISoundManager
 		}
 		return playSoundRawAt(buf, loop, volume, pos);
 	}
+
 	void stopSound(int sound)
 	{
 		maintain();
 		deleteSound(sound);
 	}
+
+	void fadeSound(int soundid, float step, float gain)
+	{
+		m_sounds_fading[soundid] = FadeState(step, getSoundGain(soundid), gain);
+	}
+
+	void doFades(float dtime)
+	{
+		m_fade_delay += dtime;
+
+		if (m_fade_delay < 0.2f)
+			return;
+
+		float chkGain = 0;
+		for (UNORDERED_MAP<int, FadeState>::iterator i = m_sounds_fading.begin();
+				i != m_sounds_fading.end();) {
+			if (i->second.step < 0.f)
+				chkGain = -(i->second.current_gain);
+			else
+				chkGain = i->second.current_gain;
+
+			if (chkGain < i->second.target_gain) {
+				i->second.current_gain += (i->second.step * m_fade_delay);
+				i->second.current_gain = rangelim(i->second.current_gain, 0, 1);
+
+				updateSoundGain(i->first, i->second.current_gain);
+				++i;
+			} else {
+				if (i->second.target_gain <= 0.f)
+					stopSound(i->first);
+
+				m_sounds_fading.erase(i++);
```

Comment:
Just pointing out: with `std::unordered_map` (so in C++11), this is completely correct only with C++14. With C++11, it will not crash, however there is no guarantee that all sounds will be updated once and only once. As this is not very bad, and classic compilers will have the standard library required by C++14 anyway, I don't think that's a problem -- however, this is something that need to be kept in mind when this iteration is more important (I'm wondering if we're not using such a pattern elsewhere in the codebase anyway).